### PR TITLE
Explicitly cache X509v3 extensions in libssl

### PR DESCRIPTION
**Pull request by mattcaswell**
A by product of processing the X509v3 extensions in an X509 object results in a SHA1 hash being calculated and stored for later use. Lots of X509 functions end up in the implicit processing of the X509v3 extensions and could result in this SHA1 hash being calculated. Prior to this PR the SHA1 hash would just use the default libctx and property query. However libssl needs to be able to specify a non-default libctx. Rather than modifying large numbers of X509 functions to add an OPENSSL_CTX parameter, we expose a new function to explicitly process the X509v3 extensions. This can be called up-front before other X509 functions are called.

### CODE REVIEW

File: crypto/x509/v3_purp.c [hunk 1]
```diff
@@ -361,21 +358,25 @@ static void x509v3_cache_extensions(X509 *x)
     EXTENDED_KEY_USAGE *extusage;
     X509_EXTENSION *ex;
     int i;
+    EVP_MD *sha1;
 
 #ifdef tsan_ld_acq
     /* fast lock-free check, see end of the function for details. */
     if (tsan_ld_acq((TSAN_QUALIFIER int *)&x->ex_cached))
-        return;
+        return (x->ex_flags & EXFLAG_INVALID) == 0;
```
Comment (richsalz) [inline crypto/x509/v3_purp.c]: I wonder if calling this bit out separately as an atomic-int field is worthwhile.

Comment (mattcaswell) [reply crypto/x509/v3_purp.c]: The `ex_flags` can be obtained by application code using `X509_get_extension_flags()`, so for backwards compat reasons we'd still have to keep the `EXFLAG_INVALID` flag anyway. If we were designing this from scratch then what you say might make sense - but probably not now.

Comment (richsalz) [reply crypto/x509/v3_purp.c]: You can keep the API, just merge both fields as the return from the "get-flags" method.

I am slightly concerned about the thread-safety of non-atomic (bitwise) operations, but oh well.

Comment (mattcaswell) [reply crypto/x509/v3_purp.c]: > You can keep the API, just merge both fields as the return from the "get-flags" method.

We could do, but I don't think its worth it.

**Comment by mattcaswell** [comment]
I rebased this to pick up the latest libcrypto.num changes.

Still looking for a reviewer for this. Ping!

**Comment by kaduk** [comment]
[fixed up approval labels per @t8m's approval]

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**